File: bootstrapping.py
import numpy as np
from numpy.typing import NDArray
from numba import njit
from time import perf_counter, time


def f(x: NDArray, p: NDArray) -> NDArray:
    return p[0] + p[1]*x + np.random.randn(*x.shape)*0.1*x


def model(x: NDArray, p: NDArray) -> NDArray:
    p[0] + p[1]*x


def fit(x: NDArray, y: NDArray) -> NDArray:
    X = np.stack((np.ones_like(x), x), axis=1)
    return np.linalg.lstsq(X, y)[0]


def bootstrap(
    x: NDArray,
    y: NDArray,
    num_iters: int = 1000,
) -> NDArray:
    p_boot = np.zeros((num_iters, 2))

    for i in range(num_iters):
        idx_boot = np.random.randint(low=0, high=x.shape[0], size=x.shape[0])
        x_boot = x[idx_boot]
        y_boot = y[idx_boot]

        p = fit(x_boot, y_boot)

        p_boot[i, :] = p
    return p_boot


if __name__ == "__main__":
    x = np.arange(0, 10, 0.01)
    p_true = np.array([0, 1])
    y = f(x, p_true)

    start = perf_counter()
    _ = bootstrap(x, y, num_iters=100_000)
    stop = perf_counter()
    print(f"Python: {(stop-start)*1000:.2f} ms")

    bootstrap = njit(bootstrap)
    fit = njit(fit)

    # warm-up
    p_boot = bootstrap(x, y, num_iters=10)

    start = perf_counter()
    p_boot = bootstrap(x, y, num_iters=100_000)
    stop = perf_counter()
    print(f"numba:  {(stop-start)*1000:.2f} ms")

    print("mean:", np.mean(p_boot, axis=0))
    print("std:", np.std(p_boot, axis=0))

    print(np.percentile(p_boot[:,0], 2.5))
    print(np.percentile(p_boot[:,0], 97.5))

    print(np.percentile(p_boot[:,1], 2.5))
    print(np.percentile(p_boot[:,1], 97.5))

File: bootstrapping.jl
using Printf
using Random
using StatsBase
using LinearAlgebra
using BenchmarkTools

f(x, p) = p[1] .+ p[2]*x .+ randn(size(x))*0.1.*x

x = collect(0:0.01:10-0.01)
p_true = (0, 1)
y = f(x, p_true);

model(x, p) = p[1] .+ p[2]*x

function fit(
    x::Vector{Float64},
    y::Vector{Float64}
)::Vector{Float64}
    X = hcat(ones(size(x)), x)
    return X \ y
end

fit(x, y)

println("Threads ", Threads.nthreads())

function bootstrap(
    x::Vector{Float64},
    y::Vector{Float64};
    num_iters::Integer = 1000
)::Matrix{Float64}
    p_boot = zeros(num_iters, 2)
    size_x = size(x)
    Threads.@threads for i in 1:num_iters
        idx_boot = sample(range(1, size_x[1]), size_x)
        x_boot = x[idx_boot]
        y_boot = y[idx_boot]

        p = fit(x_boot, y_boot)

        p_boot[i, :] = p
    end
    return p_boot
end

# warm-up
p_boot = bootstrap(x, y; num_iters=10);

b = @benchmark begin
    p_boot = bootstrap(x, y; num_iters=100_000)
end

println(b)

@printf "mean: %.3f %.3f\n" mean(p_boot, dims=1)[1] mean(p_boot, dims=1)[2]
@printf "std:  %.3f %.3f\n" std(p_boot, dims=1)[1] std(p_boot, dims=1)[2]

@printf "percentile p_1: %.3f %.3f\n" percentile(p_boot[:,1], 2.5) percentile(p_boot[:,1], 97.5)

@printf "percentile p_2: %.3f %.3f\n" percentile(p_boot[:,2], 2.5) percentile(p_boot[:,2], 97.5)
